Avoid unnecessary status update when there is no corresponding mirror pod

Random-Liu opened the pull request:
Fixes https://github.com/kubernetes/kubernetes/issues/32191.

This PR changes status manager to skip update when there is no mirror pod for a static pod.
We need this because:
1) When static pod terminates and mirror pod is deleted, this will avoid extra `syncPod`.
2) During mirror pod creation and recreation, this will avoid unnecessary `syncPod`.

Mark P1 to match the original issue.

@wojtek-t @yujuhong 
/cc @kubernetes/sig-node

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32232)

<!-- Reviewable:end -->

Random-Liu committed:
Avoid StatusManager.syncPod when there is no corresponding mirror pod for static pod.

timstclair commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -372,6 +372,10 @@ func (m *manager) syncBatch() {
 			syncedUID := uid
 			if mirrorUID, ok := podToMirror[uid]; ok {
+				if mirrorUID == "" {
+					glog.V(5).Infof("Static pod %q does not have a corresponding mirror pod; skipping", uid)
```
nit: log the name & namespace too (from `status`)

timstclair commented on pkg/kubelet/status/status_manager_test.go:
```diff
@@ -517,21 +529,22 @@ func TestStaticPodStatus(t *testing.T) {
 	verifyActions(t, m.kubeClient, []core.Action{
 		core.GetActionImpl{ActionImpl: core.ActionImpl{Verb: "get", Resource: unversioned.GroupVersionResource{Resource: "pods"}}},
 	})
+
```
nit: extra line

timstclair commented:
Just 2 nits

Random-Liu commented:
@timstclair Addressed comment.

timstclair commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -372,6 +372,10 @@ func (m *manager) syncBatch() {
 			syncedUID := uid
 			if mirrorUID, ok := podToMirror[uid]; ok {
+				if mirrorUID == "" {
+					glog.V(5).Infof("Static pod %q (%s/%s) does not have a corresponding mirror pod; skipping", uid, status.podName, status.podNamespace)
```
nit: follow the same format used for logging elsewhere: https://github.com/kubernetes/kubernetes/blob/7c27a2686ff70a11bda20c27c9cc8dc154183357/pkg/kubelet/util/format/pod.go#L34

Better yet, add a helper function in that package that takes a the name, namespace, and UID.

Random-Liu replied:
ACK.

timstclair commented:
I don't want to block this on a logging nit. If you'd like to follow up in another PR go for it.

dchen1107 commented on pkg/kubelet/pod/pod_manager.go:
```diff
@@ -237,6 +236,17 @@ func (pm *basicManager) GetUIDTranslations() (podToMirror, mirrorToPod map[types
 
 	podToMirror = make(map[types.UID]types.UID, len(pm.translationByUID))
 	mirrorToPod = make(map[types.UID]types.UID, len(pm.translationByUID))
+	// Insert empty translation mapping for all static pods.
```
Can you update this with more comment based on our offline discussion? Without the detail reason & logic behind, I am concerning the maintainability.

Random-Liu replied:
ACK.

dchen1107 commented:
The issue only affects the test infra on the efficiency, not affect any production environment. So it shouldn't blocker 1.4 release. But since the fix is relatively small with low risk, we are ok to cherrypick this to 1.4 release. Thanks!

LGTM

Random-Liu commented:
Addressed comments @dchen1107 @timstclair 
Self apply LGTM based on https://github.com/kubernetes/kubernetes/pull/32232#issuecomment-245465712 and https://github.com/kubernetes/kubernetes/pull/32232#issuecomment-245466624.

pwittrock commented:
1.  Please provide a risk assessment in the description

2. How critical is this?  blocking / important / best effort?  We have dozens of cherrypick requests and each one creates a lot of overhead because large batches introduce a high-risk of destabilizing the tests in a way that is hard to track down.   We have to break them up into smaller batches and monitor the tests to catch issues early on and track them back to their root cause.
